File: typescript/src/scan/scan.ts
import { readdir, readFile, stat } from "node:fs/promises";
import path from "node:path";

import { newSnapshot, type RepoFile, type Snapshot } from "../repo/repo.js";

const maxFileBytes = 1 << 20;
const skippedDirectories = new Set([
  ".git",
  "node_modules",
  ".venv",
  ".mypy_cache",
  ".pytest_cache",
  ".ruff_cache",
  ".stryker-tmp",
  "__pycache__",
  "dist",
  "coverage"
]);

export async function scanRepo(root: string): Promise<Snapshot> {
  const absoluteRoot = path.resolve(root);
  const files: RepoFile[] = [];
  await walk(absoluteRoot, absoluteRoot, files);
  return newSnapshot(absoluteRoot, files);
}

async function walk(root: string, current: string, files: RepoFile[]): Promise<void> {
  const entries = await readdir(current, { withFileTypes: true });
  for (const entry of entries) {
    const filePath = path.join(current, entry.name);
    if (entry.isDirectory()) {
      if (!skippedDirectories.has(entry.name)) {
        await walk(root, filePath, files);
      }
      continue;
    }
    if (!entry.isFile()) {
      continue;
    }
    const file = await readSmallTextFile(root, filePath);
    if (file !== undefined) {
      files.push(file);
    }
  }
}

async function readSmallTextFile(root: string, filePath: string): Promise<RepoFile | undefined> {
  const info = await stat(filePath);
  if (info.size > maxFileBytes) {
    return undefined;
  }
  const content = await readFile(filePath, "utf8");
  if (content.includes("\0")) {
    return undefined;
  }
  return {
    path: path.relative(root, filePath).split(path.sep).join("/"),
    content
  };
}

File: python/src/slophammer_py/scan.py
"""Repository walk producing a snapshot of small text files."""

from __future__ import annotations

import os
from pathlib import Path

from .repo import RepoFile, Snapshot, new_snapshot

MAX_FILE_BYTES = 1 << 20
SKIPPED_DIRECTORIES = {
    ".git",
    "node_modules",
    ".venv",
    ".mypy_cache",
    ".pytest_cache",
    ".ruff_cache",
    ".stryker-tmp",
    "mutants",
    "__pycache__",
    "dist",
    "coverage",
    "target",
}


def scan_repo(root: str) -> Snapshot:
    absolute_root = Path(root).resolve()
    files: list[RepoFile] = []
    for current, directories, names in os.walk(absolute_root):
        directories[:] = sorted(name for name in directories if name not in SKIPPED_DIRECTORIES)
        for name in sorted(names):
            file = read_small_text_file(absolute_root, Path(current) / name)
            if file is not None:
                files.append(file)
    return new_snapshot(str(absolute_root), files)


def read_small_text_file(root: Path, file_path: Path) -> RepoFile | None:
    try:
        if not file_path.is_file() or file_path.stat().st_size > MAX_FILE_BYTES:
            return None
        content = file_path.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError):
        return None
    if "\0" in content:
        return None
    return RepoFile(path=file_path.relative_to(root).as_posix(), content=content)
